elrond-wasm/src/storage/mappers/bi_di_mapper.rs

```diff
@@ -0,0 +1,266 @@
+use core::marker::PhantomData;
+
+use elrond_codec::{
+    multi_encode_iter_or_handle_err, EncodeErrorHandler, NestedDecode, NestedEncode, TopDecode,
+    TopEncode, TopEncodeMulti, TopEncodeMultiOutput,
+};
+
+use super::{StorageMapper, UnorderedSetMapper};
+use crate::{
+    abi::{TypeAbi, TypeName},
+    api::StorageMapperApi,
+    storage::{storage_get, storage_set, StorageKey},
+    storage_clear,
+    types::{ManagedAddress, ManagedType, ManagedVec, ManagedVecItem, MultiValueEncoded},
+};
+
+const VALUE_SUFIX: &[u8] = b"_value";
+const ID_SUFIX: &[u8] = b"_id";
+const VALUE_TO_ID_SUFFIX: &[u8] = b"_value_to_id";
+const ID_TO_VALUE_SUFFIX: &[u8] = b"_id_to_value";
+
+/// A bi-directional map, from values to ids and viceversa.
+/// The mapper is based on UnorderedSetMapper, reason why the remove is done by swap_remove
+pub struct BiDiMapper<SA, K, V>
+where
+    SA: StorageMapperApi,
+    K: TopEncode
+        + TopDecode
+        + NestedEncode
+        + NestedDecode
+        + 'static
+        + Default
+        + PartialEq
+        + ManagedVecItem,
+    V: TopEncode
+        + TopDecode
+        + NestedEncode
+        + NestedDecode
+        + 'static
+        + Default
+        + PartialEq
+        + ManagedVecItem,
+{
+    _phantom_api: PhantomData<SA>,
+    id_set_mapper: UnorderedSetMapper<SA, K>,
+    value_set_mapper: UnorderedSetMapper<SA, V>,
+    base_key: StorageKey<SA>,
+}
+
+impl<SA, K, V> StorageMapper<SA> for BiDiMapper<SA, K, V>
+where
+    SA: StorageMapperApi,
+    K: TopEncode
+        + TopDecode
+        + NestedEncode
+        + NestedDecode
+        + 'static
+        + Default
+        + PartialEq
+        + ManagedVecItem,
+    V: TopEncode
+        + TopDecode
+        + NestedEncode
+        + NestedDecode
+        + 'static
+        + Default
+        + PartialEq
+        + ManagedVecItem,
+{
+    fn new(base_key: StorageKey<SA>) -> Self {
+        let mut id_key = base_key.clone();
+        id_key.append_bytes(ID_SUFIX);
+
+        let mut value_key = base_key.clone();
+        value_key.append_bytes(VALUE_SUFIX);
+        BiDiMapper {
+            _phantom_api: PhantomData,
+            id_set_mapper: UnorderedSetMapper::<SA, K>::new(id_key),
+            value_set_mapper: UnorderedSetMapper::<SA, V>::new(value_key),
+            base_key,
+        }
+    }
+}
+
+impl<SA, K, V> BiDiMapper<SA, K, V>
+where
+    SA: StorageMapperApi,
+    K: TopEncode
+        + TopDecode
+        + NestedEncode
+        + NestedDecode
+        + 'static
+        + Default
+        + PartialEq
+        + ManagedVecItem,
```

Comment:
ManagedVecItem is not necessary, we should yield an iterator, not push it into a ManagedVec.
If we provide a method to extract all keys/values directly to a ManagedVec, the trait restriction should be only around that method.